src/puppetlabs/jdbc_util/pool.clj

```diff
@@ -0,0 +1,115 @@
+  (case option
+    :username (.setUsername config value)
+    :password (.setPassword config value)
+    :data-source-class-name (.setDataSourceClassName config value)
+    :jdbc-url (.setJdbcUrl config value)
+    :driver-class-name (.setDriverClassName config value)
+    :auto-commit (.setAutoCommit config value)
+    :connection-timeout (.setConnectionTimeout config value)
+    :idle-timeout (.setIdleTimeout config value)
+    :max-lifetime (.setMaxLifetime config value)
+    :connection-test-query (.setConnectionTestQuery config value)
+    :minimum-idle (.setMinimumIdle config value)
+    :maximum-pool-size (.setMaximumPoolSize config value)
+    :metric-registry (.setMetricRegistry config value)
+    :health-check-registry (.setHealthCheckRegistry config value)
+    :pool-name (.setPoolName config value)
+    (throw (IllegalArgumentException. (format "%s is not a supported HikariCP option" (str option))))))
+
+(defn options->hikari-config
+  [options]
+  (let [config (HikariConfig.)]
+    (doseq [[option value] options]
+      (set-option config option value))
+    (.validate config)
+    config))
+
+(defn spec->hikari-options
+  [db-spec]
+  (-> db-spec
+      (set/rename-keys {:user :username
+                        :classname :driver-class-name})
+      (assoc :jdbc-url (str "jdbc:"
+                            (:subprotocol db-spec) ":"
+                            (:subname db-spec)))
+      (dissoc :subprotocol :subname)))
+
+(defprotocol PoolStatus
+  (status [this] "Get a map representing the status of a connection pool."))
+
+(defn wrap-with-delayed-init
+  "Wraps a connection pool that loops trying to get a connection, and then runs
+  init-fn (with the connection as argument) before returning any connections to
+  the application. Accepts a timeout in ms that's used when deferencing the
+  future and by the status check. The datasource should have
+  initialization-fail-fast set before being created or this is pointless."
+  [^HikariDataSource datasource init-fn timeout]
+  (when-not (.getHealthCheckRegistry datasource)
+    (.addHealthCheckProperty datasource "connectivityCheckTimeoutMs" (str timeout))
+    (.setHealthCheckRegistry datasource (HealthCheckRegistry.)))
+  (let [init-result (atom nil)
+        pool-future
+        (future
+          (loop []
+            (if-let [result
+                     (try
+                       (with-open [conn (.getConnection datasource)]
+                         (try (init-fn conn)
+                              (catch Exception e
+                                (swap! init-result (constantly e))
+                                (log/errorf e "%s - An error was encountered during initialization." (.getPoolName datasource))))
+                         datasource)
+                       (catch SQLTransientConnectionException e
+                         (log/warnf e "%s - Error while attempting to connect to database, retrying." (.getPoolName datasource))))]
```

Comment:
I'm not wild about understanding this code hinging on knowing the return value of the log/levelf family of functions, though it's not too hard to infer because otherwise it wouldn't ever do what it's supposed to. What do you think?